### ISSUE TITLE
Android Build fails with "Failed loading Type for clojure/core"

### ISSUE THREAD

**Issue by Polyrhythm** [arcadia-unity/Arcadia#185]
Building to a Samsung Galaxy S6 on Android 6.0.1. Building the same project to desktop works no problem.

Here's a log dump: The first line about not finding the audio plugin is a separate issue and not fatal.

DLLs in the build directory are all there:
![](http://i.imgur.com/TBMZpb6.png)

```
10-25 14:09:43.523: E/Unity(22909): Unable to find libaudioplugingvrunity
10-25 14:09:43.573: D/Unity(22909):  GL_EXT_debug_marker GL_ARM_rgba8 GL_ARM_mali_shader_binary GL_OES_depth24 GL_OES_depth_texture GL_OES_depth_texture_cube_map GL_OES_packed_depth_stencil GL_OES_rgb8_rgba8 GL_EXT_read_format_bgra GL_OES_compressed_paletted_texture GL_OES_compressed_ETC1_RGB8_texture GL_OES_standard_derivatives GL_OES_EGL_image GL_OES_EGL_image_external GL_OES_EGL_sync GL_OES_texture_npot GL_OES_vertex_half_float GL_OES_required_internalformat GL_OES_vertex_array_object GL_OES_mapbuffer GL_EXT_texture_format_BGRA8888 GL_EXT_texture_rg GL_EXT_texture_type_2_10_10_10_REV GL_OES_fbo_render_mipmap GL_OES_element_index_uint GL_EXT_shadow_samplers GL_OES_texture_compression_astc GL_KHR_texture_compression_astc_ldr GL_KHR_texture_compression_astc_hdr GL_KHR_debug GL_EXT_occlusion_query_boolean GL_EXT_disjoint_timer_query GL_EXT_blend_minmax GL_EXT_discard_framebuffer GL_OES_get_program_binary GL_OES_texture_3D GL_EXT_texture_storage GL_EXT_multisampled_render_to_texture GL_OES_surfaceless_context GL_OES_texture_stencil8 GL_EXT_shade
10-25 14:09:43.573: D/Unity(22909): r_pixel_local_storage GL_ARM_shader_framebuffer_fetch GL_ARM_shader_framebuffer_fetch_depth_stencil GL_ARM_mali_program_binary GL_EXT_sRGB GL_EXT_sRGB_write_control GL_EXT_texture_sRGB_decode GL_KHR_blend_equation_advanced GL_KHR_blend_equation_advanced_coherent GL_OES_texture_storage_multisample_2d_array GL_OES_shader_image_atomic GL_EXT_robustness GL_EXT_draw_buffers_indexed GL_OES_draw_buffers_indexed GL_EXT_texture_border_clamp GL_OES_texture_border_clamp GL_EXT_texture_cube_map_array GL_OES_texture_cube_map_array GL_OES_sample_variables GL_OES_sample_shading GL_OES_shader_multisample_interpolation GL_EXT_shader_io_blocks GL_OES_shader_io_blocks GL_EXT_tessellation_shader GL_OES_tessellation_shader GL_EXT_primitive_bounding_box GL_OES_primitive_bounding_box GL_EXT_geometry_shader GL_OES_geometry_shader GL_ANDROID_extension_pack_es31a GL_EXT_gpu_shader5 GL_OES_gpu_shader5 GL_EXT_texture_buffer GL_OES_texture_buffer GL_EXT_copy_image GL_OES_copy_image GL_EXT_color_buffer_half_float GL_EXT_color_buffer_floa
10-25 14:09:43.573: D/Unity(22909): t GL_OVR_multiview GL_OVR_multiview2 GL_OVR_multiview_multisampled_render_to_texture
10-25 14:09:43.823: W/Unity(22909): OnLevelWasLoaded was found on OnLevelWasLoadedHook
10-25 14:09:43.823: W/Unity(22909): This message has been deprecated and will be removed in a later version of Unity.
10-25 14:09:43.823: W/Unity(22909): Add a delegate to SceneManager.sceneLoaded instead to get notifications after scene loading has completed
10-25 14:09:43.823: W/Unity(22909):  
10-25 14:09:43.823: W/Unity(22909): (Filename:  Line: 375)
10-25 14:09:45.223: I/Unity(22909): AssemblyLoadException: Failed loading Init Type for clojure/core
10-25 14:09:45.223: I/Unity(22909):   at clojure.lang.Compiler.TryLoadInitType (System.String relativePath) [0x00000] in <filename unknown>:0 
10-25 14:09:45.223: I/Unity(22909):   at clojure.lang.RT.load (System.String relativePath, Boolean failIfNotFound) [0x00000] in <filename unknown>:0 
10-25 14:09:45.223: I/Unity(22909):   at clojure.lang.RT.load (System.String relativePath) [0x00000] in <filename unknown>:0 
10-25 14:09:45.223: I/Unity(22909):   at clojure.lang.RT.DoInit () [0x00000] in <filename unknown>:0 
10-25 14:09:45.223: I/Unity(22909):   at clojure.lang.RT..cctor () [0x00000] in <filename unknown>:0 
10-25 14:09:45.223: I/Unity(22909): Rethrow as TypeInitializationException: An exception was thrown by the type initializer for clojure.lang.RT
10-25 14:09:45.223: I/Unity(22909):   at ArcadiaBehaviour.Awake () [0x00000] in <filename unknown>:0 
10-25 14:09:45.223: I/Unity(22909):  
10-25 14:09:45.223: I/Unity(22909): (Filename:  Line: -1)
10-25 14:09:45.403: I/Unity(22909): AssemblyLoadException: Failed loading Init Type for clojure/core
10-25 14:09:45.403: I/Unity(22909):   at clojure.lang.Compiler.TryLoadInitType (System.String relativePath) [0x00000] in <filename unknown>:0 
10-25 14:09:45.403: I/Unity(22909):   at clojure.lang.RT.load (System.String relativePath, Boolean failIfNotFound) [0x00000] in <filename unknown>:0 
10-25 14:09:45.403: I/Unity(22909):   at clojure.lang.RT.load (System.String relativePath) [0x00000] in <filename unknown>:0 
10-25 14:09:45.403: I/Unity(22909):   at clojure.lang.RT.DoInit () [0x00000] in <filename unknown>:0 
10-25 14:09:45.403: I/Unity(22909):   at clojure.lang.RT..cctor () [0x00000] in <filename unknown>:0 
10-25 14:09:45.403: I/Unity(22909): Rethrow as TypeInitializationException: An exception was thrown by the type initializer for clojure.lang.RT
10-25 14:09:45.403: I/Unity(22909):   at ArcadiaBehaviour.Awake () [0x00000] in <filename unknown>:0 
10-25 14:09:45.403: I/Unity(22909):  
10-25 14:09:45.403: I/Unity(22909): (Filename:  Line: -1)
```

**Comment by nasser** [comment]
57ab756fce46968941c40b89ba2eb9954758260c didnt fix it

**Comment by selfsame** [comment]
Some notes on this issue:

A working android build can be achieved by: 
* manually AOTing `clojure.core` and user namespaces into `Assets` 
* disabling the current android build script in `Arcadia/Editor/BuildPostProcess.cs`

the Arcadia Build script has the following issues:
* doesn't know how to sign the default `unsigned (Debug)` setting in `Player Settings > Publishing Settings`
* doesn't sign the newer android SDK versions correctly (possibly a setting with `jarsigner`, `apksigner` was signing them fine)
* has the code commented out to swap the signed+aligned apk for the original
* **doesn't seem to make a correct apk**, all the arcadia/clojure/user dlls are there but not found by Unity at runtime, resulting in `FileNotFoundException: Could not locate clojure.core.cljc.dll or clojure/core.cljc on load path.`
  * diffing a working manual aot apk & build script apk found some differences in the `assets/bin/Data/data.unity3d` file that might be references to the dll files.

Sneaking DLLs into the apk seems quite brittle, as the android sdk tooling is subject to change. One possible resolution to this issue is remove the android build script, and add instructions to the documentation on how to manual AOT before building.

**Comment by nasser** [comment]
This should be fixed by the `build` branch, merged into `develop` as of 8cbd0c3c831eb7df5a70bbe3a119e482b8dac41c. Reopen if it resurfaces!
